### ISSUE TITLE
Release-Acquire chains

### ISSUE THREAD

**Issue by hernanponcedeleon** [paulmckrcu/litmus#11]
The following program has a potential related-acq pair (`P0 -> P2`) that can be broken by an intervening thread (`P1`) with a relaxed instruction
```
❯ cat weird-C.litmus
C weird
{}

P0(intptr_t *x, intptr_t *y)
{
	WRITE_ONCE(*x, 1);
	atomic_set_release(y, 1);
}

P1(intptr_t *x, intptr_t *y)
{
	int r0 = atomic_fetch_add_relaxed(2, y); 
}

P2(intptr_t *x, intptr_t *y)
{
    int r0 = atomic_read_acquire(y);
    int r1 = READ_ONCE(*x);
}

exists (2:r0=3 /\ 2:r1=0)
```
The LKMM from [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/memory-model/linux-kernel.cat) allows the chain to be broken
```
❯ herd7 -model ../lkmm/tools/memory-model/linux-kernel.cat -macros ../lkmm/tools/memory-model/linux-kernel.def -bell ../lkmm/tools/memory-model/linux-kernel.bell weird-C.litmus -I ../lkmm/tools/memory-model/
Test weird Allowed
States 7
2:r0=0; 2:r1=0;
2:r0=0; 2:r1=1;
2:r0=1; 2:r1=1;
2:r0=2; 2:r1=0;
2:r0=2; 2:r1=1;
2:r0=3; 2:r1=0;
2:r0=3; 2:r1=1;
Ok
Witnesses
Positive: 1 Negative: 9
Condition exists (2:r0=3 /\ 2:r1=0)
Observation weird Sometimes 1 9
Time weird 0.01
Hash=f3a36a6d5313d0bcee071c2a06783c41
```
The chain could be fixed by making the RMW in `P1` a release
```
❯ cat weird-C-fixed.litmus
C weird
{}

P0(intptr_t *x, intptr_t *y)
{
    WRITE_ONCE(*x, 1);
    atomic_set_release(y, 1);
}

P1(intptr_t *x, intptr_t *y)
{
    int r0 = atomic_fetch_add_release(2, y);
}

P2(intptr_t *x, intptr_t *y)
{
    int r0 = atomic_read_acquire(y);
    int r1 = READ_ONCE(*x);
}

exists (2:r0=3 /\ 2:r1=0)

❯ herd7 -model ../lkmm/tools/memory-model/linux-kernel.cat -macros ../lkmm/tools/memory-model/linux-kernel.def -bell ../lkmm/tools/memory-model/linux-kernel.bell weird-C-fixed.litmus -I ../lkmm/tools/memory-model/
Test weird Allowed
States 6
2:r0=0; 2:r1=0;
2:r0=0; 2:r1=1;
2:r0=1; 2:r1=1;
2:r0=2; 2:r1=0;
2:r0=2; 2:r1=1;
2:r0=3; 2:r1=1;
No
Witnesses
Positive: 0 Negative: 9
Condition exists (2:r0=3 /\ 2:r1=0)
Observation weird Never 0 9
Time weird 0.01
Hash=2d1ad2a445f148efe7634b8992363e5d
```

This behaviour would neither be observed in an ARM
```
❯ cat weird-ARM.litmus
AArch64 weird
{
0:X1=x; 0:X3=y;
1:X1=x; 1:X3=y;
2:X1=x; 2:X3=y;
}
P0           | P1                | P2;
MOV  W0,#1   | MOV   W0,#2       | LDAR W0, [X3];
STR  W0,[X1] | LDXR  W2,[X3]     | LDR  W2, [X1];
STLR W0,[X3] | ADD   W2,W2,W0    |;
             | STXR  W1,W2,[X3]  |;
 
exists
(2:X0=3 /\ 2:X2=0)
❯ herd7 -model cat/aarch64.cat weird-ARM.litmus
Test weird Allowed
States 6
2:X0=0; 2:X2=0;
2:X0=0; 2:X2=1;
2:X0=1; 2:X2=1;
2:X0=2; 2:X2=0;
2:X0=2; 2:X2=1;
2:X0=3; 2:X2=1;
No
Witnesses
Positive: 0 Negative: 15
Condition exists (2:X0=3 /\ 2:X2=0)
Observation weird Never 0 15
Time weird 0.01
Hash=342601047a6d69dee710e4e94d28e4fe
```
or PowerPC machine
```
❯ cat weird-PPC.litmus
PPC weird
{
0:r2=x; 0:r4=y;
1:r2=x; 1:r4=y;
2:r2=x; 2:r4=y;
}
P0            | P1			| P2		;
li   r1,1     | lwz  r3,0(r4)		| lwz r1,0(r4)	;
stw  r1,0(r2) | cmpw r3,r3		| cmpw r3,r3	;
stw  r1,0(r4) | beq  LC00		| beq  LC01	;
 	      | LC00:			| LC01: 	;
 	      | addi r3,r3,1		| isync		;
 	      | stw r3,0(r4) 		| lwz r3, 0(r2)	;
 
exists
(2:r1=3 /\ 2:r3=0)
❯ herd7 -model cat/power.cat weird-PPC.litmus
Test weird Allowed
States 6
2:r1=0; 2:r3=0;
2:r1=0; 2:r3=1;
2:r1=1; 2:r3=0;
2:r1=1; 2:r3=1;
2:r1=2; 2:r3=0;
2:r1=2; 2:r3=1;
No
Witnesses
Positive: 0 Negative: 18
Condition exists (2:r1=3 /\ 2:r3=0)
Observation weird Never 0 18
Time weird 0.01
Hash=09d4d22cf368195d3c18a8c9acf34a2c
```

My question is then why the LKMM allows this behaviour? Is there any weaker architecture supported by the kernel (e.g. Alpha) where this behaviour would be observable?

**Comment by paulmckrcu** [comment]
On Fri, Apr 29, 2022 at 03:27:49AM -0700, Zeta wrote:
> The following program has a potential related-acq pair (`P0 -> P2`) that can be broken by an intervening thread (`P1`) with a relaxed instruction
> ```
> ❯ cat weird-C.litmus
> C weird
> {}
> 
> P0(intptr_t *x, intptr_t *y)
> {
> 	WRITE_ONCE(*x, 1);
> 	atomic_set_release(y, 1);
> }
> 
> P1(intptr_t *x, intptr_t *y)
> {
> 	int r0 = atomic_fetch_add_relaxed(2, y); 
> }
> 
> P2(intptr_t *x, intptr_t *y)
> {
>     int r0 = atomic_read_acquire(y);
>     int r1 = READ_ONCE(*x);
> }
> 
> exists (2:r0=3 /\ 2:r1=0)
> ```
> The LKMM from [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/memory-model/linux-kernel.cat) allows the chain to be broken
> ```
> ❯ herd7 -model ../lkmm/tools/memory-model/linux-kernel.cat -macros ../lkmm/tools/memory-model/linux-kernel.def -bell ../lkmm/tools/memory-model/linux-kernel.bell weird-C.litmus -I ../lkmm/tools/memory-model/
> Test weird Allowed
> States 7
> 2:r0=0; 2:r1=0;
> 2:r0=0; 2:r1=1;
> 2:r0=1; 2:r1=1;
> 2:r0=2; 2:r1=0;
> 2:r0=2; 2:r1=1;
> 2:r0=3; 2:r1=0;
> 2:r0=3; 2:r1=1;
> Ok
> Witnesses
> Positive: 1 Negative: 9
> Condition exists (2:r0=3 /\ 2:r1=0)
> Observation weird Sometimes 1 9
> Time weird 0.01
> Hash=f3a36a6d5313d0bcee071c2a06783c41
> ```
> The chain could be fixed by making the RMW in `P1` a release
> ```
> ❯ cat weird-C-fixed.litmus
> C weird
> {}
> 
> P0(intptr_t *x, intptr_t *y)
> {
>     WRITE_ONCE(*x, 1);
>     atomic_set_release(y, 1);
> }
> 
> P1(intptr_t *x, intptr_t *y)
> {
>     int r0 = atomic_fetch_add_release(2, y);
> }
> 
> P2(intptr_t *x, intptr_t *y)
> {
>     int r0 = atomic_read_acquire(y);
>     int r1 = READ_ONCE(*x);
> }
> 
> exists (2:r0=3 /\ 2:r1=0)
> 
> ❯ herd7 -model ../lkmm/tools/memory-model/linux-kernel.cat -macros ../lkmm/tools/memory-model/linux-kernel.def -bell ../lkmm/tools/memory-model/linux-kernel.bell weird-C-fixed.litmus -I ../lkmm/tools/memory-model/
> Test weird Allowed
> States 6
> 2:r0=0; 2:r1=0;
> 2:r0=0; 2:r1=1;
> 2:r0=1; 2:r1=1;
> 2:r0=2; 2:r1=0;
> 2:r0=2; 2:r1=1;
> 2:r0=3; 2:r1=1;
> No
> Witnesses
> Positive: 0 Negative: 9
> Condition exists (2:r0=3 /\ 2:r1=0)
> Observation weird Never 0 9
> Time weird 0.01
> Hash=2d1ad2a445f148efe7634b8992363e5d
> ```
> 
> This behaviour would neither be observed in an ARM
> ```
> ❯ cat weird-ARM.litmus
> AArch64 weird
> {
> 0:X1=x; 0:X3=y;
> 1:X1=x; 1:X3=y;
> 2:X1=x; 2:X3=y;
> }
> P0           | P1                | P2;
> MOV  W0,#1   | MOV   W0,#2       | LDAR W0, [X3];
> STR  W0,[X1] | LDXR  W2,[X3]     | LDR  W2, [X1];
> STLR W0,[X3] | ADD   W2,W2,W0    |;
>              | STXR  W1,W2,[X3]  |;
>  
> exists
> (2:X0=3 /\ 2:X2=0)
> ❯ herd7 -model cat/aarch64.cat weird-ARM.litmus
> Test weird Allowed
> States 6
> 2:X0=0; 2:X2=0;
> 2:X0=0; 2:X2=1;
> 2:X0=1; 2:X2=1;
> 2:X0=2; 2:X2=0;
> 2:X0=2; 2:X2=1;
> 2:X0=3; 2:X2=1;
> No
> Witnesses
> Positive: 0 Negative: 15
> Condition exists (2:X0=3 /\ 2:X2=0)
> Observation weird Never 0 15
> Time weird 0.01
> Hash=342601047a6d69dee710e4e94d28e4fe
> ```
> or PowerPC machine
> ```
> ❯ cat weird-PPC.litmus
> PPC weird
> {
> 0:r2=x; 0:r4=y;
> 1:r2=x; 1:r4=y;
> 2:r2=x; 2:r4=y;
> }
> P0            | P1			| P2		;
> li   r1,1     | lwz  r3,0(r4)		| lwz r1,0(r4)	;
> stw  r1,0(r2) | cmpw r3,r3		| cmpw r3,r3	;
> stw  r1,0(r4) | beq  LC00		| beq  LC01	;
>  	      | LC00:			| LC01: 	;
>  	      | addi r3,r3,1		| isync		;
>  	      | stw r3,0(r4) 		| lwz r3, 0(r2)	;
>  
> exists
> (2:r1=3 /\ 2:r3=0)
> ❯ herd7 -model cat/power.cat weird-PPC.litmus
> Test weird Allowed
> States 6
> 2:r1=0; 2:r3=0;
> 2:r1=0; 2:r3=1;
> 2:r1=1; 2:r3=0;
> 2:r1=1; 2:r3=1;
> 2:r1=2; 2:r3=0;
> 2:r1=2; 2:r3=1;
> No
> Witnesses
> Positive: 0 Negative: 18
> Condition exists (2:r1=3 /\ 2:r3=0)
> Observation weird Never 0 18
> Time weird 0.01
> Hash=09d4d22cf368195d3c18a8c9acf34a2c
> ```
> 
> My question is then why the LKMM allows this behaviour? Is there any weaker architecture supported by the kernel (e.g. Alpha) where this behaviour would be observable?

If I understand and remember correctly, this is a side effect of LKMM
ignoring the 2+2W litmus test.  We ignored this one because it seems to
be an attractive nuisance, in other words, anything you could do with
2+2W can be done better some other way.

Adding Alan Stern on CC for his perspective.

							Thanx, Paul

**Comment by paulmckrcu** [comment]
On Mon, May 02, 2022 at 09:55:00AM -0700, Paul E. McKenney wrote:
> On Fri, Apr 29, 2022 at 03:27:49AM -0700, Zeta wrote:
> > The following program has a potential related-acq pair (`P0 -> P2`) that can be broken by an intervening thread (`P1`) with a relaxed instruction
> > ```
> > ❯ cat weird-C.litmus
> > C weird
> > {}
> > 
> > P0(intptr_t *x, intptr_t *y)
> > {
> > 	WRITE_ONCE(*x, 1);
> > 	atomic_set_release(y, 1);
> > }
> > 
> > P1(intptr_t *x, intptr_t *y)
> > {
> > 	int r0 = atomic_fetch_add_relaxed(2, y); 
> > }
> > 
> > P2(intptr_t *x, intptr_t *y)
> > {
> >     int r0 = atomic_read_acquire(y);
> >     int r1 = READ_ONCE(*x);
> > }
> > 
> > exists (2:r0=3 /\ 2:r1=0)
> > ```
> > The LKMM from [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/memory-model/linux-kernel.cat) allows the chain to be broken
> > ```
> > ❯ herd7 -model ../lkmm/tools/memory-model/linux-kernel.cat -macros ../lkmm/tools/memory-model/linux-kernel.def -bell ../lkmm/tools/memory-model/linux-kernel.bell weird-C.litmus -I ../lkmm/tools/memory-model/
> > Test weird Allowed
> > States 7
> > 2:r0=0; 2:r1=0;
> > 2:r0=0; 2:r1=1;
> > 2:r0=1; 2:r1=1;
> > 2:r0=2; 2:r1=0;
> > 2:r0=2; 2:r1=1;
> > 2:r0=3; 2:r1=0;
> > 2:r0=3; 2:r1=1;
> > Ok
> > Witnesses
> > Positive: 1 Negative: 9
> > Condition exists (2:r0=3 /\ 2:r1=0)
> > Observation weird Sometimes 1 9
> > Time weird 0.01
> > Hash=f3a36a6d5313d0bcee071c2a06783c41
> > ```
> > The chain could be fixed by making the RMW in `P1` a release
> > ```
> > ❯ cat weird-C-fixed.litmus
> > C weird
> > {}
> > 
> > P0(intptr_t *x, intptr_t *y)
> > {
> >     WRITE_ONCE(*x, 1);
> >     atomic_set_release(y, 1);
> > }
> > 
> > P1(intptr_t *x, intptr_t *y)
> > {
> >     int r0 = atomic_fetch_add_release(2, y);
> > }
> > 
> > P2(intptr_t *x, intptr_t *y)
> > {
> >     int r0 = atomic_read_acquire(y);
> >     int r1 = READ_ONCE(*x);
> > }
> > 
> > exists (2:r0=3 /\ 2:r1=0)
> > 
> > ❯ herd7 -model ../lkmm/tools/memory-model/linux-kernel.cat -macros ../lkmm/tools/memory-model/linux-kernel.def -bell ../lkmm/tools/memory-model/linux-kernel.bell weird-C-fixed.litmus -I ../lkmm/tools/memory-model/
> > Test weird Allowed
> > States 6
> > 2:r0=0; 2:r1=0;
> > 2:r0=0; 2:r1=1;
> > 2:r0=1; 2:r1=1;
> > 2:r0=2; 2:r1=0;
> > 2:r0=2; 2:r1=1;
> > 2:r0=3; 2:r1=1;
> > No
> > Witnesses
> > Positive: 0 Negative: 9
> > Condition exists (2:r0=3 /\ 2:r1=0)
> > Observation weird Never 0 9
> > Time weird 0.01
> > Hash=2d1ad2a445f148efe7634b8992363e5d
> > ```
> > 
> > This behaviour would neither be observed in an ARM
> > ```
> > ❯ cat weird-ARM.litmus
> > AArch64 weird
> > {
> > 0:X1=x; 0:X3=y;
> > 1:X1=x; 1:X3=y;
> > 2:X1=x; 2:X3=y;
> > }
> > P0           | P1                | P2;
> > MOV  W0,#1   | MOV   W0,#2       | LDAR W0, [X3];
> > STR  W0,[X1] | LDXR  W2,[X3]     | LDR  W2, [X1];
> > STLR W0,[X3] | ADD   W2,W2,W0    |;
> >              | STXR  W1,W2,[X3]  |;
> >  
> > exists
> > (2:X0=3 /\ 2:X2=0)
> > ❯ herd7 -model cat/aarch64.cat weird-ARM.litmus
> > Test weird Allowed
> > States 6
> > 2:X0=0; 2:X2=0;
> > 2:X0=0; 2:X2=1;
> > 2:X0=1; 2:X2=1;
> > 2:X0=2; 2:X2=0;
> > 2:X0=2; 2:X2=1;
> > 2:X0=3; 2:X2=1;
> > No
> > Witnesses
> > Positive: 0 Negative: 15
> > Condition exists (2:X0=3 /\ 2:X2=0)
> > Observation weird Never 0 15
> > Time weird 0.01
> > Hash=342601047a6d69dee710e4e94d28e4fe
> > ```
> > or PowerPC machine
> > ```
> > ❯ cat weird-PPC.litmus
> > PPC weird
> > {
> > 0:r2=x; 0:r4=y;
> > 1:r2=x; 1:r4=y;
> > 2:r2=x; 2:r4=y;
> > }
> > P0            | P1			| P2		;
> > li   r1,1     | lwz  r3,0(r4)		| lwz r1,0(r4)	;
> > stw  r1,0(r2) | cmpw r3,r3		| cmpw r3,r3	;
> > stw  r1,0(r4) | beq  LC00		| beq  LC01	;
> >  	      | LC00:			| LC01: 	;
> >  	      | addi r3,r3,1		| isync		;
> >  	      | stw r3,0(r4) 		| lwz r3, 0(r2)	;
> >  
> > exists
> > (2:r1=3 /\ 2:r3=0)
> > ❯ herd7 -model cat/power.cat weird-PPC.litmus
> > Test weird Allowed
> > States 6
> > 2:r1=0; 2:r3=0;
> > 2:r1=0; 2:r3=1;
> > 2:r1=1; 2:r3=0;
> > 2:r1=1; 2:r3=1;
> > 2:r1=2; 2:r3=0;
> > 2:r1=2; 2:r3=1;
> > No
> > Witnesses
> > Positive: 0 Negative: 18
> > Condition exists (2:r1=3 /\ 2:r3=0)
> > Observation weird Never 0 18
> > Time weird 0.01
> > Hash=09d4d22cf368195d3c18a8c9acf34a2c
> > ```
> > 
> > My question is then why the LKMM allows this behaviour? Is there any weaker architecture supported by the kernel (e.g. Alpha) where this behaviour would be observable?
> 
> If I understand and remember correctly, this is a side effect of LKMM
> ignoring the 2+2W litmus test.  We ignored this one because it seems to
> be an attractive nuisance, in other words, anything you could do with
> 2+2W can be done better some other way.
> 
> Adding Alan Stern on CC for his perspective.

Well, for one thing the PPC-weird litmus test is wrong.  The code for P1 
contains an unnecessary read-to-write barrier and the addition 
operation isn't atomic -- although I guess being non-atomic doesn't 
matter here.  More importantly, the addition adds 1 to y instead of 2 
(you can see this in the output from herd7; 2:r1 takes on the values 0, 
1, and 2 but never 3).

As for why the LKMM doesn't require ordering for this test...  I do not 
believe this is related to 2+2W.  Think instead in terms of the LKMM's 
operational model:

	The store-release in P0 means that the x=1 write will propagate 
	to each CPU before the y=1 write does.

	Since y=3 at the end, we know that y=1 (and hence x=1 too)
	propagates to P1 before the addition occurs.  And we know that 
	y=3 propagates to P2 before the load-acquire executes.

	But we _don't_ know that either y=1 or x=1 propagates to P2 
	before y=3 does!  If the store in P1 were a store-release then 
	this would have to be true (as you saw in your tests), but it 
	isn't.

In other words, the litmus test could execute with the following 
temporal ordering:

	P0		P1		P2
	----------	---------	----------
	Write x=1
	Write y=1

		[x=1 and y=1 propagate to P1]

			Read y=1
			Write y=3

		[y=3 propagates to P2]

					Read y=3
					Read x=0

		[x=1 and y=1 propagate to P2]

(Note that when y=1 propagates to P2, it doesn't do anything because it 
won't overwrite the coherence-later store y=3.)

It may be true that none of the architectures supported by Linux will 
allow this outcome for the test (although I suspect that the PPC-weird 
version _would_ be allowed if you fixed it).  At any rate, disallowing 
this result in the memory model would probably require more effort than 
would be worthwhile.

Alan

**Comment by hernanponcedeleon** [comment]
Thanks @paulmck and @AlanStern for the answer. There were indeed several problems with the PPC litmus (sorry for this)
- we should add 2 instead of 1
- there was a missing `lwsync` in `P0` for the release semantics
- in `P2` the fake dependency should use `r1` instead of `r3`
- access are not atomic in `P1`. While `herd7` should support the atomic instructions I could not make it run with those. However, as Alan mentioned, this should not change anything in this example

> P1 contains an unnecessary read-to-write barrier 

I did not understand this comment. Is it about the fake dependency? I was using the compiler mapping (modulo the atomic accesses) in Figure 9 in [this paper](https://plv.mpi-sws.org/scfix/paper.pdf).

With the first 3 from above fixed (hopefully there are no more errors), the behaviour is not observable in power
```
❯ cat weird-PPC.litmus
PPC weird
 {
 0:r2=x; 0:r4=y;
 1:r2=x; 1:r4=y;
 2:r2=x; 2:r4=y;
 }
 P0             | P1            | P2		;
 li   r1,1      | lwz  r3,0(r4)	| lwz r1,0(r4)	;
 stw  r1,0(r2)  | cmpw r3,r3    | cmpw r1,r1	;
 lwsync         | beq  LC00     | beq  LC01	;
 stw  r1,0(r4)  | LC00:         | LC01: 	;
                | addi r3,r3,2  | isync		;
                | stw r3,0(r4)  | lwz r3, 0(r2)	;

 exists
 (2:r1=3 /\ 2:r3=0)
❯ herd7 -model cat/power.cat weird-PPC.litmus
Test weird Allowed
States 6
2:r1=0; 2:r3=0;
2:r1=0; 2:r3=1;
2:r1=1; 2:r3=1;
2:r1=2; 2:r3=0;
2:r1=2; 2:r3=1;
2:r1=3; 2:r3=1;
No
Witnesses
Positive: 0 Negative: 14
Condition exists (2:r1=3 /\ 2:r3=0)
Observation weird Never 0 14
Time weird 0.02
Hash=8a3f95437cb2f5d124004e7e06dc286e
```

> disallowing this result in the memory model would probably require more effort than
would be worthwhile.

From the point of view of the formalisation of the memory model (and assuming indeed no architecture actually allows this behaviour), the first idea that would come to my mind is to use similar relations / axioms than the ones in the hardware memory models which forbids this behaviour. Do you think that adding those would result in removing other behaviours that are indeed intended to be accepted by LKMM? Is this what you mean by "it requires more effort than worthwhile"?

**Comment by AlanStern** [comment]
On Tue, May 03, 2022 at 12:42:25AM -0700, Hernan Ponce de Leon wrote:
> Thanks @paulmck and @AlanStern for the answer. There were indeed several problems with the PPC litmus (sorry for this)
> - we should add 2 instead of 1
> - there was a missing `lwsync` in `P0` for the release semantics
> - in `P2` the fake dependency should use `r1` instead of `r3`
> - access are not atomic in `P1`. While `herd7` should support the atomic instructions I could not make it run with those. However, as Alan mentioned, this should not change anything in this example
> 
> > P1 contains an unnecessary read-to-write barrier 
> 
> I did not understand this comment. Is it about the fake dependency? I 
> was using the compiler mapping (modulo the atomic accesses) in Figure 
> 9 in [this paper](https://plv.mpi-sws.org/scfix/paper.pdf).

Yes, I was referring to the fake dependency.  It serves no purpose 
because the code contains a genuine data dependency anyway.

> With the first 3 from above fixed (hopefully there are no more errors), the behaviour is not observable in power
> ```
> ❯ cat weird-PPC.litmus
> PPC weird
>  {
>  0:r2=x; 0:r4=y;
>  1:r2=x; 1:r4=y;
>  2:r2=x; 2:r4=y;
>  }
>  P0             | P1            | P2		;
>  li   r1,1      | lwz  r3,0(r4)	| lwz r1,0(r4)	;
>  stw  r1,0(r2)  | cmpw r3,r3    | cmpw r1,r1	;
>  lwsync         | beq  LC00     | beq  LC01	;
>  stw  r1,0(r4)  | LC00:         | LC01: 	;
>                 | addi r3,r3,2  | isync		;
>                 | stw r3,0(r4)  | lwz r3, 0(r2)	;
> 
>  exists
>  (2:r1=3 /\ 2:r3=0)
> ❯ herd7 -model cat/power.cat weird-PPC.litmus
> Test weird Allowed
> States 6
> 2:r1=0; 2:r3=0;
> 2:r1=0; 2:r3=1;
> 2:r1=1; 2:r3=1;
> 2:r1=2; 2:r3=0;
> 2:r1=2; 2:r3=1;
> 2:r1=3; 2:r3=1;
> No
> Witnesses
> Positive: 0 Negative: 14
> Condition exists (2:r1=3 /\ 2:r3=0)
> Observation weird Never 0 14
> Time weird 0.02
> Hash=8a3f95437cb2f5d124004e7e06dc286e
> ```

Okay, this seems reasonable.  I tried running this litmus test through 
Peter Sewell's PPCMEM memory model, which I think is a more accurate 
approximation to actual PPC behavior than the herd model is.  It also 
says the test is not allowed.

The reason is because barriers in PPCMEM are B-cumulative.  This means 
that since the lwsync barrier in P0 propagates to P1 before the store in 
P1 executes, the barrier has to propagate to P2 before the store does.  
In fact, your litmus test is a simple variant of the ISA2 test, which is 
a standard example showing how B-cumulativity works.

The LKMM does not include B-cumulativity; that's why it allows this 
test.  We judged that putting B-cumulativity into the model would make 
it unnecessarily complicated.

> > disallowing this result in the memory model would probably require more effort than
> would be worthwhile.
> 
> From the point of view of the formalisation of the memory model (and 
> assuming indeed no architecture actually allows this behaviour), the 
> first idea that would come to my mind is to use similar relations / 
> axioms than the ones in the hardware memory models which forbids this 
> behaviour. Do you think that adding those would result in removing 
> other behaviours that are indeed intended to be accepted by LKMM? Is 

Yes, unless it was done extremely carefully.

> this what you mean by "it requires more effort than worthwhile"?

In part.  I also meant that the new axioms might not be compatible with 
the underlying ideas behind the LKMM (although to be fair, we have 
already added one or two axioms like that).

Alan

**Comment by hernanponcedeleon** [comment]
> Okay, this seems reasonable. I tried running this litmus test through Peter Sewell's PPCMEM memory model, which I think is a more accurate approximation to actual PPC behavior than the herd model is. 

Thanks for the pointer to PPCMEM. I was not aware of this one and seems like a nice resource.

> The LKMM does not include B-cumulativity; that's why it allows this test. We judged that putting B-cumulativity into the model would make it unnecessarily complicated.

Ok, I think this answer my original question so I'll close the issue.

Thanks @AlanStern and @paulmck for the explanations.

**Comment by paulmckrcu** [comment]
On Wed, May 04, 2022 at 12:59:06AM -0700, Hernan Ponce de Leon wrote:
> > Okay, this seems reasonable. I tried running this litmus test through Peter Sewell's PPCMEM memory model, which I think is a more accurate approximation to actual PPC behavior than the herd model is. 
> 
> Thanks for the pointer to PPCMEM. I was not aware of this one and seems like a nice resource.
> 
> > The LKMM does not include B-cumulativity; that's why it allows this test. We judged that putting B-cumulativity into the model would make it unnecessarily complicated.
> 
> Ok, I think this answer my original question so I'll close the issue.
> 
> Thanks @AlanStern and @paulmck for the explanations.

Looking forward to seeing what you come up with!

							Thanx, Paul
